### PR TITLE
SI-9823 Collections perf: favor virtual call over instanceof

### PR DESCRIPTION
**Pull request by retronym**
This avoids concurrent usages of collections in NUMA architectures
from falling of a performance cliff due to an implementation detail
of interface instanceof checks in HotSpot JVM.

The VM contains a one element cache in the metadata of each class
to record the most recent successful test for a interface.

For example:

```
classOf[Serializable].isAssignableFrom(classOf[Some[_]])
```

Would, in addition to returning `true`, set:

```
classOf[Some[_]]._secondary_super_cache = classOf[Serializable]
```

This is done to hide the fact that interface tests are O(N),
where N is the number of inherited interfaces.

This scheme is discussed in "Fast Subtype Checking for the
HotSpot JVM" (Click, Rose) [1]

However, if this cache repeatedly misses, not only are we
exposed to the linear search of the secondary super type array,
but we are also required to write back to the cache. If other
cores are operating on the same cache line, this can lead to
a significant slowdown ("cache thrashing"). This effect will
by most (or only?) visible on multi socket machines.

The pathological case is:

```
scala> Iterator.continually(List(classOf[Product], classOf[Serializable])).flatten.take(100).map(intf => intf.isAssignableFrom(classOf[Some[_]])).count(x => x)
res19: Int = 100
```

Which, if run on a multi-socket machine, should be much slower
than:

```
scala> (Iterator.continually(classOf[Product]).take(50) ++ Iterator.continually(classOf[Serializable]).take(50)).map(intf => intf.isAssignableFrom(classOf[Some[_]])).count(x => x)
res20: Int = 100
```

This commit avoids a interface test in a hot path in the collections
by instead using virtual dispatch to differentiate between
IndexedSeqLike and other collections. HotSpot will still use some
shared bookkeeping ("inline cache" [2]) at the callsites of
this method, but these stabilize in the megamorphic usage and no
longer force expensive cache line synchronization.

[1] https://www.researchgate.net/publication/221552851_Fast_subtype_checking_in_the_HotSpot_JVM
[2] https://wiki.openjdk.java.net/display/HotSpot/PerformanceTechniques

### CODE REVIEW

**Comment by retronym** [comment]
Review by @sjrd @Ichoran 

This is an an alternative version of #5351 that has a smaller delta to the collections API, by just adding `sizeHint` that is `-1` for non cheap sizes.

**Comment by adriaanm** [comment]
I'm not a performance expert, but LGTM on the correctness front.

File: src/library/scala/collection/GenTraversableOnce.scala [hunk 1]
```diff
@@ -96,6 +96,12 @@ trait GenTraversableOnce[+A] extends Any {
    */
   def size: Int
 
+   /** The size of this $coll if it is can be cheaply computed
+   *
+   *  @return    the number of elements in this $coll, or -1 if the size cannot be determined cheaply
+   */
+  protected[collection] def sizeHint: Int = -1
```
Comment (sjrd) [inline src/library/scala/collection/GenTraversableOnce.scala]: I'm not sure about restricting the visibility of this method. Supposedly, if I create my own type of collection, I might want to call `sizeHint` on another collection? Or should I only ever call the methods `sizeHint(coll)` & co.?

Comment (sjrd) [inline src/library/scala/collection/GenTraversableOnce.scala]: Btw, using `sizeHint` for the name of this method doesn't look like a good idea to me:
- It does not have the same meaning as the other methods called `sizeHint` (which are _also_ in the hierarchy): this one is a "getter" of sorts, while the others are "setters".
- It creates an overload between a `()`-less `def` and other defs _with_ `()`. I've always found this was not very clean (and, IIRC, it's not even allowed in dotty anymore)

Can we find another name?

Comment (adriaanm) [reply src/library/scala/collection/GenTraversableOnce.scala]: `protected def sizeIfCheap`?

**Comment by soc** [comment]
`sizeInfo`? Assuming we don't want to introduce some kind of more sophisticated design as in psp?

File: src/library/scala/collection/GenTraversableOnce.scala (hunk 1 above)
Comment (soc) [reply src/library/scala/collection/GenTraversableOnce.scala]: `sizeInfo`? Assuming we don't want to introduce some kind of more sophisticated design as in psp?

Comment (sjrd) [reply src/library/scala/collection/GenTraversableOnce.scala]: > `protected def sizeIfCheap`

Sounds good to me.

@soc `sizeInfo` doesn't convey the "optionality" or "inaccuracy" of what it does, IMO.

**Comment by retronym** [comment]
`sizeHintOf` ?

File: src/library/scala/collection/GenTraversableOnce.scala (hunk 1 above)
Comment (soc) [reply src/library/scala/collection/GenTraversableOnce.scala]: I think it has done a good job in this use-case: https://github.com/non/psp-view/blob/master/core/src/main/scala/SizeInfo.scala#L8

But yeah, `sizeIfCheap` is fine, at least we can add something more advanced with `sizeInfo` later.

Comment (retronym) [reply src/library/scala/collection/GenTraversableOnce.scala]: > I'm not sure about restricting the visibility of this method.

I was mostly being conservative about offering this for public consumption when it is designed to be used in the collection/builder protocol.

**Comment by adriaanm** [comment]
Let's go with `protected def sizeIfCheap: Int`, since there isn't much more information that we can convey with an `Int` anyway (i.e., it's either equal to `size` or a marker to denote `size` is not "cheap" -- happy to pick another adjective, but I can't think of any and I don't think it's terribly crucial since it's internal).

**Comment by retronym** [comment]
`sizeIfCheap` it is, then. I've left it `protected[collection]` for now (it needs to be accessible from within `Builder#sizeHint(TraversableLike)`, so protected is insufficient. @sjrd was asking whether it should in fact be public. While it would be okay as a public method, I'm hesitant to do this without a good name.

**Comment by adriaanm** [comment]
LGTM

**Comment by sjrd** [comment]
LGTM as well.

File: src/library/scala/collection/GenTraversableOnce.scala [hunk 2]
```diff
@@ -96,6 +96,12 @@ trait GenTraversableOnce[+A] extends Any {
    */
   def size: Int
 
+   /** The size of this $coll if it is can be cheaply computed
```
Comment (analytically) [inline src/library/scala/collection/GenTraversableOnce.scala:99]: Small typo in the comment.

Comment (SethTisue) [reply src/library/scala/collection/GenTraversableOnce.scala:99]: @analytically thanks, I'll get it in my next batch of typo fixes

**Comment by retronym** [comment]
The underlying performance problem in the JVM is tracked as [JDK-8180450](https://bugs.openjdk.java.net/browse/JDK-8180450)

**Comment by He-Pin** [comment]
optimization in netty: https://github.com/netty/netty/pull/12709

**Pull request merged**